contrib/imdocker/imdocker.c

```diff
@@ -0,0 +1,1809 @@
+
+/* utility functions */
+static CURLcode docker_get(imdocker_req_t *req, const char* url);
+static char* dupDockerContainerName(const char* pname);
+static rsRetVal SubmitMsg(docker_cont_logs_inst_t *pInst, docker_cont_logs_buf_t *pBufData,
+		const uchar* pszTag);
+/* support multi-line */
+static rsRetVal
+SubmitMsg2(docker_cont_logs_inst_t *pInst, docker_cont_logs_buf_t *pBufData, const uchar* pszTag);
+static size_t imdocker_container_list_curlCB(void *data, size_t size, size_t nmemb, void *buffer);
+static size_t imdocker_container_logs_curlCB(void *data, size_t size, size_t nmemb, void *buffer);
+static sbool get_stream_info(const uchar* data, size_t size, int8_t *stream_type, size_t *payload_size);
+static int8_t is_valid_stream_type(int8_t stream_type);
+
+/* unit tests */
+#ifdef ENABLE_IMDOCKER_UNIT_TESTS
+static void UnitTests_imdocker_run(void);
+#endif
+
+/* Module static data */
+DEF_IMOD_STATIC_DATA
+DEFobjCurrIf(glbl)
+DEFobjCurrIf(prop)
+DEFobjCurrIf(parser)
+DEFobjCurrIf(datetime)
+DEFobjCurrIf(statsobj)
+DEFobjCurrIf(ruleset)
+
+statsobj_t *modStats;
+STATSCOUNTER_DEF(ctrSubmit, mutCtrSubmit)
+STATSCOUNTER_DEF(ctrLostRatelimit, mutCtrLostRatelimit)
+STATSCOUNTER_DEF(ctrCurlError, mutctrCurlError)
+
+const char* DFLT_dockerAPIUnixSockAddr  = "/var/run/docker.sock";
+const char* DFLT_dockerAPIAdd           = "http://localhost:2375";
+const char* DFLT_apiVersionStr          = "v1.27";
+const char* DFLT_listContainersOptions  = "";
+const char* DFLT_getContainerLogOptions = "timestamps=0&follow=1&stdout=1&stderr=1&tail=1";
+const char* DFLT_getContainerLogOptionsWithoutTail = "timestamps=0&follow=1&stdout=1&stderr=1";
+
+/* Overall module configuration structure here. */
+struct modConfData_s {
+	rsconf_t *pConf;  /* our overall config object */
+	uchar    *apiVersionStr;
+	uchar    *listContainersOptions;
+	uchar    *getContainerLogOptions;
+	uchar    *getContainerLogOptionsWithoutTail;
+	int      iPollInterval;  /* in seconds */
+	uchar    *dockerApiUnixSockAddr;
+	uchar    *dockerApiAddr;
+	sbool    retrieveNewLogsFromStart;
+	int      containersLimit;
+	int      trimLineOverBytes;
+};
+
+static modConfData_t *loadModConf = NULL;
+static modConfData_t *runModConf = NULL;
+
+static prop_t *pInputName = NULL;   /* our inputName currently is always "imdocker", and this will hold it */
+static prop_t *pLocalHostIP = NULL; /* a pseudo-constant propterty for 127.0.0.1 */
+
+static ratelimit_t *ratelimiter = NULL;
+
+/* module-global parameters */
+static struct cnfparamdescr modpdescr[] = {
+	{ "apiversionstr", eCmdHdlrString, 0 },
+	{ "dockerapiunixsockaddr", eCmdHdlrString, 0 },
+	{ "dockerapiaddr", eCmdHdlrString, 0 },
+	{ "listcontainersoptions", eCmdHdlrString, 0 },
+	{ "getcontainerlogoptions", eCmdHdlrString, 0 },
+	{ "pollinginterval", eCmdHdlrPositiveInt, 0 },
+	{ "retrievenewlogsfromstart", eCmdHdlrBinary, 0 },
+	{ "trimlineoverbytes", eCmdHdlrPositiveInt, 0 },
+	{ "ruleset", eCmdHdlrString, 0 }
```

Comment:
parameter "ruleset" is not implemented

Reply:
yes, it ingests all containers. I think each container needs to be identified using template via container name.  i'm testing something along those lines. will upload soon.

Reply:
if we support "ruleset" is not a question of how much it ingests. It makes user's life easier if it does. I came across it when I wanted to split imdocker messages from the rest of the message stream inside the test. Nevertheless, it's not a hard requirement to support multiple rulesets. But than that parameter needs to be removed from the param block as it really does not exist.

Reply:
"if the module does support ruleset parameter" I wanted to say ;-)

Reply:
> > It's expected that there are many containers running, there is metadata about which container the message is from that is the key to making things work. This metadata then needs to be translated into container name, etc.
> 
> I see. So the module must read **everything** from docker and then sort it in buckets? If so, it's a bit hard to test, but at least we can find ways to do it:
> 
> 1. generate a unique identifier for each test
> 2. make the plugin only output data with that identifier
> 3. force `make check` to not run these tests in parallel (it's unfortunate, but we have other places where we need to do that)
> 
> Am I on the right path? If so, what would be the best method to do 2.?

I have an commit ready which illustrates what David mentioned. 
the idea is to run the container using **--name** option, then output the log line along with the name in order to do a proper count.
should i send it up?